spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/outbound/MongoDbOutboundGateway.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.mongodb.outbound;
+
+import org.bson.Document;
+
+import org.springframework.data.mongodb.MongoDbFactory;
+import org.springframework.data.mongodb.core.MongoOperations;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.convert.DefaultDbRefResolver;
+import org.springframework.data.mongodb.core.convert.MappingMongoConverter;
+import org.springframework.data.mongodb.core.convert.MongoConverter;
+import org.springframework.data.mongodb.core.mapping.MongoMappingContext;
+import org.springframework.data.mongodb.core.query.BasicQuery;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.expression.TypeLocator;
+import org.springframework.expression.spel.support.StandardTypeLocator;
+import org.springframework.integration.expression.ExpressionUtils;
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+/**
+ * Makes outbound operations to query a MongoDb database using a {@link MongoOperations}
+ *
+ * @author Xavier Padró
+ * @since 5.0
+ */
+public class MongoDbOutboundGateway extends AbstractReplyProducingMessageHandler {
+
+	private volatile EvaluationContext evaluationContext;
+
+	private volatile Expression queryExpression;
+
+	private volatile MongoOperations mongoTemplate;
+
+	private volatile MongoDbFactory mongoDbFactory;
+
+	private volatile MongoConverter mongoConverter;
+
+	private volatile boolean expectSingleResult = false;
+
+	private volatile Class<?> entityClass = Document.class;
+
+	private volatile Expression collectionNameExpression;
+
+	public MongoDbOutboundGateway(MongoDbFactory mongoDbFactory) {
+		this(mongoDbFactory, new MappingMongoConverter(new DefaultDbRefResolver(mongoDbFactory),
+				new MongoMappingContext()));
+	}
+
+	public MongoDbOutboundGateway(MongoDbFactory mongoDbFactory, MongoConverter mongoConverter) {
+		Assert.notNull(mongoDbFactory, "mongoDbFactory must not be null.");
+		Assert.notNull(mongoConverter, "mongoConverter must not be null.");
+		this.mongoDbFactory = mongoDbFactory;
+		this.mongoConverter = mongoConverter;
+	}
+
+	public MongoDbOutboundGateway(MongoOperations mongoTemplate) {
+		Assert.notNull(mongoTemplate, "mongoTemplate must not be null.");
+		this.mongoTemplate = mongoTemplate;
+	}
+
+	public void setQueryExpression(Expression queryExpression) {
+		this.queryExpression = queryExpression;
```

Comment:
`Assert.notNull()` - just for fail fast, since we really don't support `null` for this property.